// @cnips/simplelog/dist/index.js
import util from "util";
export const levels = {
    debug: -1,
    info: 0,
    warn: 1,
    error: 2,
};
/**
 * Logger class for managing application logs with different levels.
 * Provides methods to log messages at various levels and retrieve logs.
 *
 * example usage:
 * ```typescript
 * import { Logger } from "@cnips/logger";
 * const logger = new Logger("MyApp", Logger.levels.debug);
 * logger.info("This is an info message");
 * logger.warn("This is a warning message");
 * logger.error("This is an error message");
 * logger.debug("This is a debug message");
 * ```
 * @remarks
 * You can use `printf` syntax in the log messages, e.g.:
 * ```typescript
 * logger.info("User %s logged in at %s", username, new Date().toISOString());
 * ```
 * This will format the message with the provided arguments.
 */
export class Logger {
    constructor(app, level = levels.info) {
        this.app = app;
        this.level = level;
        this.logs = [];
    }
    /**
     * Logs an informational message.
     *
     * @remarks
     * This method formats the message with any additional arguments and prepends a timestamp and application name.
     * The message is only logged if the current log level is set to `info` or lower.
     * @param message
     * @param args
     */
    info(message, ...args) {
        if (levels.info >= this.level) {
            message = util.format(message, ...args);
            let msg = `${new Date().toISOString()} [${this.app}] INFO ${message}`;
            this.logs.push(msg);
        }
    }
    warn(message, ...args) {
        if (levels.warn >= this.level) {
            message = util.format(message, ...args);
            let msg = `${new Date().toISOString()} [${this.app}] WARN ${message}`;
            this.logs.push(msg);
        }
    }
    error(message, ...args) {
        if (levels.error >= this.level) {
            message = util.format(message, ...args);
            let msg = `${new Date().toISOString()} [${this.app}] ERROR ${message}`;
            this.logs.push(msg);
        }
    }
    debug(message, ...args) {
        if (levels.debug >= this.level) {
            message = util.format(message, ...args);
            let msg = `${new Date().toISOString()} [${this.app}] DEBUG ${message}`;
            this.logs.push(msg);
        }
    }
    getLogs() {
        return this.logs;
    }
    clearLogs() {
        this.logs = [];
    }
    string() {
        return this.logs.join('\n');
    }
}
